src/high-level/matrix.lisp

```diff
@@ -0,0 +1,560 @@
+         (declare (dynamic-extent pos))
+         (let ((numrows (matrix-nrows matrix))
+               (numcols (matrix-ncols matrix)))
+           (declare (type alexandria:non-negative-fixnum numrows numcols))
+           (policy-cond:with-expectations (> speed safety)
+               ((assertion (valid-matrix-index-p pos numrows numcols)))
+             (let ((row (first pos))
+                   (col (second pos)))
+               (declare (type alexandria:non-negative-fixnum row col))
+               (let ((index (ecase (matrix-layout matrix)
+                              (:row-major (+ col (the fixnum (* row numcols))))
+                              (:column-major (+ row (the fixnum (* col numrows)))))))
+                 (declare (type alexandria:array-index index))
+                 (setf (aref (,storage-sym matrix) index)
+                       (coerce new-value ',type)))))))
+       
+       (defmethod into! ((function function) (matrix ,name))
+         (let ((index 0) ;; TODO: make if less iffy
+               (storage (storage matrix)))
+           (if (eql :row-major (layout matrix))
+               (loop :for j :below (nrows matrix) :do
+                 (loop :for i :below (ncols matrix) :do
+                   (setf (aref storage index)
+                         (coerce (funcall function j i) ',type))
+                   (incf index)))
+               (loop :for j :below (ncols matrix) :do
+                 (loop :for i :below (nrows matrix) :do
+                   (setf (aref storage index)
+                         (coerce (funcall function i j) ',type))
+                   (incf index))))
+           matrix)))))
+
+(defun pprint-matrix (stream matrix &optional colon-p at-sign-p)
+  "Pretty-print a matrix MATRIX to the stream STREAM."
+  (declare (ignore colon-p)
+           (ignore at-sign-p))
+  (flet ((print-real (x)
+           (format stream "~6,3f" x))
+         (print-complex (z)
+           (format stream "~6,3f ~:[+~;-~]~6,3fj"
+                   (realpart z)
+                   (minusp (imagpart z))
+                   (abs (imagpart z))))
+         (print-int (x)
+           (format stream "~3d" x)))
+    (let* ((rows (nrows matrix))
+           (cols (ncols matrix))
+           (type (element-type matrix))
+           (print-entry
+             (cond
+               ((subtypep type 'complex) #'print-complex)
+               ((subtypep type 'integer) #'print-int)
+               (t #'print-real))))
+      (pprint-logical-block (stream nil)
+        (print-unreadable-object (matrix stream :type t)
+          (format stream "(~Dx~D):" rows cols)
+          (dotimes (r rows)
+            (pprint-newline :mandatory stream)
+            (dotimes (c cols)
+              (funcall print-entry (tref matrix r c))
+              (unless (cl:= c (1- cols))
+                (write-string "    " stream)))))))))
+
+(set-pprint-dispatch 'matrix 'pprint-matrix)
+
+(defgeneric square-matrix-p (matrix)
```

Comment:
I don't necessarily mind having this be a generic, but I don't really see the point since it's not specialized for the various matrix types. As it stands it also fails on non-matrices (`(square-matrix-p (ones '(2)))` has no applicable method). Same goes for a few others in here (e.g. `identity-matrix-p`, `unitary-matrix-p` and so on).

Reply:
Agree. 837f859